### PR TITLE
INT-163 Refactor ApiClient to expose more HTTP response data

### PR DESCRIPTION
**Pull request by philippe-leonhardt-cko**
[INT-163 Refactor ApiClient to expose more HTTP response data](https://checkout.atlassian.net/browse/INT-163)

### CODE REVIEW

File: src/CheckoutSdk/HttpResponseMessageExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            return await ComposeCheckoutHttpResponseMessage<TContent>(httpResponseMessage, typeof(TContent));
+        }
+
+        /// <summary>
+        /// Converts a provided <paramref name="httpResponseMessage"/> into a <see cref="CheckoutHttpResponseMessage{dynamic}"/>.
+        /// </summary>
+        /// <param name="httpResponseMessage">The <see cref="HttpResponseMessage"/> to be transformed.</param>
+        /// <param name="responseType">The type of the response.</param>
+        /// <returns>The <see cref="CheckoutHttpResponseMessage{dynamic}"/> representation of the <see cref="HttpResponseMessage"/>.</returns>
+        public static async Task<CheckoutHttpResponseMessage<dynamic>> ConvertToChekoutHttpResponseMessage(this HttpResponseMessage httpResponseMessage, Type responseType)
```
Comment (daniel-munch-cko) [inline src/CheckoutSdk/HttpResponseMessageExtensions.cs:37]: Typo, should be `ConvertToCheckoutHttpResponseMessage `

Comment (daniel-munch-cko) [reply src/CheckoutSdk/HttpResponseMessageExtensions.cs:37]: seems to happen more often here

File: src/CheckoutSdk/ApiClient.cs [hunk 2]
```diff
@@ -122,11 +123,11 @@ public async Task<TResult> PostAsync<TResult>(string path, IApiCredentials crede
                 idempotencyKey: idempotencyKey
                 ))
             {
-                return await DeserializeJsonAsync<TResult>(httpResponse);
+                return await httpResponse.ConvertToChekoutHttpResponseMessage<TResult>();
             }
         }
 
-        public async Task<dynamic> PostAsync(string path, IApiCredentials credentials, Dictionary<HttpStatusCode, Type> resultTypeMappings, CancellationToken cancellationToken, object request = null, string idempotencyKey = null)
+        public async Task<CheckoutHttpResponseMessage<dynamic>> PostAsync(string path, IApiCredentials credentials, Dictionary<HttpStatusCode, Type> resultTypeMappings, CancellationToken cancellationToken, object request = null, string idempotencyKey = null)
```
Comment (daniel-munch-cko) [inline src/CheckoutSdk/ApiClient.cs:130]: I think this method might be the source of your test failures for `Microsoft.CSharp.RuntimeBinder.RuntimeBinderException : The best overloaded method match for 'Checkout.CheckoutHttpResponseMessage<Checkout.Payments.CaptureResponse>.CheckoutHttpResponseMessage(System.Net.HttpStatusCode, System.Net.Http.Headers.HttpResponseHeaders, Checkout.Payments.CaptureResponse)' has some invalid arguments`

there seems to be a bit too much dynamic going on here - can you point me to the method where this particular overload is used? I'm having a bit of hard time navigating in this PR, and currently don't have a VS Studio on my laptop I could use

File: src/CheckoutSdk/HttpResponseMessageExtensions.cs (hunk 1 above)
Comment (philippe-leonhardt-cko) [reply src/CheckoutSdk/HttpResponseMessageExtensions.cs:37]: fixed with commit 0fcd798

File: src/CheckoutSdk/ApiClient.cs (hunk 2 above)
Comment (philippe-leonhardt-cko) [reply src/CheckoutSdk/ApiClient.cs:130]: It's being used in 2 locations
- https://github.com/checkout/checkout-sdk-net/pull/102/files#diff-4268ee75d24bfad7856e52ec5739825f2b4d8a58919216607dfca67b4babecb0R79
- https://github.com/checkout/checkout-sdk-net/pull/102/files#diff-5c372cea63939a75da112d4449f741b0821dd313796ad4fa8958b558c509ec31R47

Comment (daniel-munch-cko) [reply src/CheckoutSdk/ApiClient.cs:130]: Thanks - I think in general this looks overly abstracted and can most likely be simplified, removing the need for `dynamic` altogether.

E.g. https://github.com/checkout/checkout-sdk-net/pull/102/files#diff-5c372cea63939a75da112d4449f741b0821dd313796ad4fa8958b558c509ec31R47

has only one single entry for the SourceResponseMappings, so I wonder why it needs a mapping at all? 

I'll give this another look a bit later, but if you can try and see if you can remove these generalisations I think it'd be a good start

**Comment by rsaestrela** [comment]
Closing due to the lack of activity. We will release a [new major version 3.0.0 of the SDK](https://github.com/checkout/checkout-sdk-net/pull/105) in the next days that will change the way the SDK is instantiated, new business features and some architectural improvements that will help to achieve this outcome, possibly in a future major version as it requires major/breaking changes.

For further questions, please feel free to open an issue.

**Pull request closed without merging**